common/src/main/java/com/wynntils/features/user/players/WynntilsCosmeticsFeature.java

```diff
@@ -0,0 +1,196 @@
+            // Cancel default cape rendering, so ours doesn't cause a double up of capes
+            event.setCanceled(true);
+        }
+    }
+
+    @SubscribeEvent
+    public void onElytraRender(PlayerRenderLayerEvent.Elytra event) {
+        if (shouldRenderCosmetic(event.getPlayer(), Type.ELYTRA)) {
+            // This might not be necessary?
+            event.setCanceled(true);
+        }
+    }
+
+    private boolean shouldRenderCosmetic(Player player, Type type) {
+        if (!Managers.Connection.onServer()) return false;
+        if (player.isInvisible() || !player.isModelPartShown(PlayerModelPart.CAPE)) return false;
+        if (McUtils.player().is(player) && !renderOwnCape) return false;
+        if (Models.Player.getUser(player.getUUID()) == null
+                || Models.Player.getUserCosmeticTexture(player.getUUID()) == null) return false;
+
+        CosmeticInfo cosmetics = Models.Player.getUser(player.getUUID()).cosmetics();
+        return (type == Type.CAPE ? cosmetics.hasCape() : cosmetics.hasElytra());
+    }
+
+    private ResourceLocation getTexture(Player player) {
+        ResourceLocation[] textures = Models.Player.getUserCosmeticTexture(player.getUUID());
+        if (textures == null) return null;
+
+        int frames = textures.length;
+        if (frames == 1) return textures[0];
+
+        // This makes animated capes cycle through all their frames in 2 seconds,
+        // regardless of how many frames of animation are present. Could be changed to an account setting?
+        double percentage = ((System.currentTimeMillis() % 2000) / 2000d);
+        int currentFrame = (int) (frames * percentage);
+        return textures[currentFrame];
+    }
+
+    private static final class WynntilsCapeLayer
+            extends RenderLayer<AbstractClientPlayer, PlayerModel<AbstractClientPlayer>> {
+        private final WynntilsCosmeticsFeature parent;
+
+        private WynntilsCapeLayer(
+                RenderLayerParent<AbstractClientPlayer, PlayerModel<AbstractClientPlayer>> renderLayerParent,
+                WynntilsCosmeticsFeature parent) {
+            super(renderLayerParent);
+            this.parent = parent;
+        }
+
+        @Override
+        public void render(
+                PoseStack poseStack,
+                MultiBufferSource buffer,
+                int packedLight,
+                AbstractClientPlayer player,
+                float limbSwing,
+                float limbSwingAmount,
+                float partialTick,
+                float ageInTicks,
+                float netHeadYaw,
+                float headPitch) {
+            if (!parent.shouldRenderCosmetic(player, Type.CAPE)) return;
+
+            ResourceLocation texture = parent.getTexture(player);
+            if (texture == null) return;
+
+            poseStack.pushPose();
+            poseStack.translate(0.0f, 0.0f, 0.125f);
+            double d = Mth.lerp(partialTick, player.xCloakO, player.xCloak)
```

Comment:
Is this math copied from MC source?

Reply:
yupp

i can try to clean it up if you want? this code is for sure a huge mess

Reply:
I think this is different from other cases, where we want to fully preserve mc code. So yes, cleanup please.